server/aws-lsp-identity/src/iam/iamProvider.ts

```diff
@@ -1,29 +1,198 @@
+            const stsClient = new STSClient({
+                region: params.profile.settings?.region || this.defaultRegion,
+                credentials: parentCredentials,
+            })
+            const assumeRoleInput: AssumeRoleCommandInput = {
+                RoleArn: params.profile.settings?.role_arn,
+                RoleSessionName: params.profile.settings?.role_session_name || `session-${Date.now()}`,
+                DurationSeconds: 3600,
+            }
+
+            // Add MFA fields to assume role request if MultiFactorAuthPresent is required
+            const mfaRequired = await checkMfaRequired(
+                parentCredentials,
+                ['sts:AssumeRole'],
+                params.profile.settings?.region
+            )
+            if (mfaRequired) {
+                const response = await this.requestMfa(params)
+                assumeRoleInput.SerialNumber = response.mfaSerial
+                assumeRoleInput.TokenCode = response.code
+
+                // Add the MFA serial number to the profile
+                const updatedProfile = {
+                    ...params.profile,
+                    settings: { ...params.profile.settings, mfa_serial: response.mfaSerial },
+                }
+                params.profileStore.save({ profiles: [updatedProfile], ssoSessions: [] })
+                // Update params.profile to ensure STS cache key is generated with updated MFA serial number
+                params.profile = updatedProfile
+            }
+
+            // Call AssumeRole API
+            const command = new AssumeRoleCommand(assumeRoleInput)
+            const { Credentials } = await stsClient.send(command)
+            if (!Credentials?.AccessKeyId || !Credentials.SecretAccessKey) {
+                throw new AwsError(
+                    'Failed to generate credentials for assumed role',
+                    AwsErrorCodes.E_CANNOT_CREATE_STS_CREDENTIAL
+                )
+            }
+            return {
+                accessKeyId: Credentials.AccessKeyId,
+                secretAccessKey: Credentials.SecretAccessKey,
+                sessionToken: Credentials.SessionToken,
+                expiration: Credentials.Expiration,
+            }
+        } catch (e) {
+            params.observability.logging.log(`Error generating STS credentials.`)
+            throw e
+        }
+    }
+
+    // Request an MFA code from the language client
+    private async requestMfa(params: IamFlowParams): Promise<GetMfaCodeResult> {
+        let timeoutId: NodeJS.Timeout | undefined
+        const timeout = new Promise<never>(
+            (_, reject) =>
+                (timeoutId = setTimeout(
+                    () => reject(new AwsError('MFA code request timed out', AwsErrorCodes.E_MFA_REQUIRED)),
+                    mfaTimeout
+                ))
+        )
+        const response = await Promise.race([
+            params.handlers.sendGetMfaCode({
```

Comment:
Does sendGetMfaCode support a cancellationToken? When you race the timeout, that lsp request will still wait for a response. With a token, you can then cancel it, and clients can then query that the request will be ignored.